compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -575,9 +575,54 @@ pub fn orphan_check(tcx: TyCtxt<'_>, impl_def_id: DefId) -> Result<(), OrphanChe
         return Ok(());
     }
 
+    let trait_ref = normalize_trait_ref(tcx, trait_ref, impl_def_id);
+
     orphan_check_trait_ref::<!>(trait_ref, InCrate::Local, |ty| Ok(ty)).unwrap()
 }
 
+fn normalize_trait_ref<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    trait_ref: ty::TraitRef<'tcx>,
+    impl_def_id: DefId,
+) -> ty::TraitRef<'tcx> {
+    let delay_bug = || {
+        tcx.sess.delay_span_bug(
+            tcx.def_span(impl_def_id),
+            format!(
+                "orphan check: failed to normalize `{trait_ref}` while checking {impl_def_id:?}"
+            ),
+        )
+    };
+
+    let infcx = tcx.infer_ctxt().build();
+    let cause = ObligationCause::dummy();
+    let param_env = tcx.param_env(impl_def_id);
+
+    if infcx.next_trait_solver() {
+        let mut fulfill_cx = <dyn TraitEngine<'_>>::new(&infcx);
+        let args = trait_ref.args.iter().map(|arg| {
+            if let ty::GenericArgKind::Type(ty) = arg.unpack()
+                && let ty::Alias(..) = ty.kind()
+            {
+                match infcx.at(&cause, param_env).structurally_normalize(ty, &mut *fulfill_cx) {
+                    Ok(ty) => ty,
+                    _ => Ty::new_error(tcx, delay_bug()),
+                }.into()
+            } else {
+                arg
+            }
+        });
+        ty::TraitRef::new(tcx, trait_ref.def_id, args)
+    } else {
+        let ocx = ObligationCtxt::new(&infcx);
+        let trait_ref = ocx.normalize(&cause, param_env, trait_ref);
+        if !ocx.select_all_or_error().is_empty() {
```

Comment:
```suggestion
        if !ocx.select_where_possible().is_empty() {
```
I think this should currently ICE on
```rust
trait Trait {
    type Assoc;
}

struct Local;
impl<T: Trait> foreign::Trait<T, T::Assoc, ()> for Local {}
```
ambiguity during normalization is alright as the associated type can be covered by something else